File: 0133-clone-graph/0133-clone-graph.js
/**
 * // Definition for a Node.
 * function Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {Node} node
 * @return {Node}
 */
var cloneGraph = function(node) {
  const copy = {};
  
  const dfs = (node) => {
    if (!node) return node;
    
    if (!(node.val in copy)) {
      copy[node.val] = new Node(node.val);
      copy[node.val].neighbors = node.neighbors.map(dfs);
    }
    
    return copy[node.val];
  }
  
  return dfs(node);
};

File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        def dfs(node):
            if not node:
                return node
            if node.val not in d:
                d[node.val] = Node(node.val)
                d[node.val].neighbors = [dfs(neighbor) for neighbor in node.neighbors]
            return d[node.val]
        
        
        d = defaultdict()
        return dfs(node)
        